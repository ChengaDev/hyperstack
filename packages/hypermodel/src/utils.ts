// Some functions taken from FoalTS

import { pbkdf2, randomBytes, timingSafeEqual } from 'crypto'
import { promisify } from 'util'

import { strictEqual } from 'assert'

export function generateHexToken(size: number) {
  return randomBytes(size).toString('hex')
}

/**
 * Legacy function to hash passwords. Only kept for backward compatibility.
 * @param password
 */
async function parsePassword(password: string): Promise<string> {
  const salt = (await promisify(randomBytes)(16)).toString('hex')
  const iterations = 100000
  const keylen = 64
  const digest = 'sha256'
  const derivedKey = await promisify(pbkdf2)(
    password,
    salt,
    iterations,
    keylen,
    digest
  )
  return `pbkdf2_${digest}$${iterations}$${salt}$${derivedKey.toString('hex')}`
}

/**
 * Hash a password using the PBKDF2 algorithm.
 *
 * Configured to use PBKDF2 + HMAC + SHA256.
 * The result is a 64 byte binary string (or hex if the legacy option is true).
 *
 * The random salt is 16 bytes long.
 * The number of iterations is 150000.
 * The length key is 32 bytes long.
 *
 * @export
 * @param {string} plainTextPassword - The password to hash.
 * @param {{ legacy?: boolean }} [options={}]
 * @returns {Promise<string>} The derived key with the algorithm name, the number of iterations and the salt.
 */
export async function hashPassword(
  plainTextPassword: string,
  options: { legacy?: boolean } = {}
): Promise<string> {
  // apply basic safeguard
  if (plainTextPassword.trim().length < 6) {
    throw new Error('Password invalid: too short')
  }

  if (options.legacy) {
    return parsePassword(plainTextPassword)
  }
  const saltBuffer = await promisify(randomBytes)(16)
  const iterations = 150000
  const keylen = 32
  const digest = 'sha256'
  const derivedKeyBuffer = await promisify(pbkdf2)(
    plainTextPassword,
    saltBuffer,
    iterations,
    keylen,
    digest
  )

  const salt = saltBuffer.toString('base64')
  const derivedKey = derivedKeyBuffer.toString('base64')
  return `pbkdf2_${digest}$${iterations}$${salt}$${derivedKey}`
}

/**
 * Compare a plain text password and a hash to see if they match.
 *
 * @export
 * @param {string} plainTextPassword - The password in clear text.
 * @param {string} passwordHash - The password hash generated by the `hashPassword` function.
 * @param {{ legacy?: boolean }} [options={}]
 * @returns {Promise<boolean>} True if the hash and the password match. False otherwise.
 */
export async function verifyPassword(
  plainTextPassword: string,
  passwordHash: string,
  options: { legacy?: boolean } = {}
): Promise<boolean> {
  const legacy = options.legacy || false
  const [algorithm, iterations, salt, derivedKey] = passwordHash.split('$')

  strictEqual(algorithm, 'pbkdf2_sha256', 'Invalid algorithm.')

  strictEqual(typeof iterations, 'string', 'Invalid password format.')
  strictEqual(typeof salt, 'string', 'Invalid password format.')
  strictEqual(typeof derivedKey, 'string', 'Invalid password format.')
  strictEqual(
    isNaN(parseInt(iterations, 10)),
    false,
    'Invalid password format.'
  )

  const saltBuffer = Buffer.from(salt, legacy ? 'hex' : 'base64')
  const derivedKeyBuffer = Buffer.from(derivedKey, legacy ? 'hex' : 'base64')
  const digest = 'sha256' // TODO: depends on the algorthim var
  const password = await promisify(pbkdf2)(
    plainTextPassword,
    legacy ? saltBuffer.toString('hex') : saltBuffer,
    parseInt(iterations, 10),
    derivedKeyBuffer.length,
    digest
  )
  return timingSafeEqual(password, derivedKeyBuffer)
}
