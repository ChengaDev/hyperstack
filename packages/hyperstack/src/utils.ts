/* eslint-disable no-console */
import fs from 'fs'
import glob from 'glob'
import {
  bold,
  cyanBright,
  gray,
  green,
  magenta,
  magentaBright,
  red,
  underline,
  white,
  whiteBright,
  yellow,
} from 'colorette'
import L from 'lodash'
import {
  asOpenAPIPath,
  getRouteMap,
  getRouteSnippets,
  routemapToOpenAPI,
} from '@hyperstackjs/hypercontroller'
import asTable from 'as-table'
import { getJobCounts } from '@hyperstackjs/hyperworker'
import type { Context } from './types'

export const sleep = (t: number) =>
  new Promise((resolve, _reject) => setTimeout(() => resolve(true), t))

export const requireModule = (f: string) => {
  let mod = null
  try {
    mod = require(f)
  } catch (e: any) {
    console.log(`error dynamically loading '${f}'\n${e.toString()}`)
    throw e
  }
  return mod.default || mod
}
export const load = (patt: string) =>
  L.sortBy(glob.sync(patt), L.identity).map(requireModule)

export const loadMap = (patt: string) =>
  L.sortBy(glob.sync(patt), L.identity).reduce((acc: any, f: string) => {
    acc[f] = requireModule(f)
    return acc
  }, {})

export const exportRoutes = (controllers: any, toFile: string) => {
  const routemap = getRouteMap(controllers)
  const spec = routemapToOpenAPI(routemap)
  fs.writeFileSync(toFile, JSON.stringify(spec.getSpec(), null, 2))
}

const snipmap: Record<string, string> = {
  wget: 'shell_wget',
  curl: 'shell_curl',
  httpie: 'shell_httpie',
}
const normalizeSnippets = (snip: string) => snipmap[snip] || snip

export const printJobCounts = async (workers: any) => {
  const counts = await getJobCounts(Object.values(workers))
  console.log(
    asTable.configure({
      print: (v, k) => {
        if (k === 'name') {
          return bold(yellow(v)).toString()
        } else if (k === 'failed') {
          return red(v).toString()
        } else if (k === 'active') {
          return yellow(v).toString()
        } else if (k === 'completed') {
          return green(v).toString()
        } else {
          return gray(v).toString()
        }
      },
    })(counts)
  )
}

export const printRoutesV2 = (controllers: any, snippets: string[] = []) => {
  const routemap = getRouteMap(controllers)

  // print the route map
  L.forEach(routemap, (mapping: any, ctl: string) => {
    console.log(`${green(ctl)} ${gray(mapping.path)}`)
    mapping.actions.forEach((action: any) => {
      action.routes.forEach((route: any) => {
        const verb = route.verb
        const path = asOpenAPIPath(route.parsedPath)
        console.log(`  ${yellow(verb.toUpperCase())}\t${whiteBright(path)}`)
        if (snippets && snippets.length > 0) {
          const snips = getRouteSnippets(
            routemap,
            verb,
            path,
            snippets.map(normalizeSnippets)
          )
          snips.snippets.forEach((snip: any) => {
            console.log(``)
            console.log(`${underline(bold(white(snip.title)))}`)
            console.log(`${gray(snip.content)}`)
            console.log(``)
          })
        }
      })
    })
    console.log('\n')
  })
}
export const printBanner = (context: Context, print: (s: string) => void) => {
  const logger = context.logger()
  const { bold, green, cyan, underline } = logger.colors
  const stat = (s: any) => underline(bold(green(s)))
  const setting = (s: any) => bold(cyan(s))
  const store = context.store()
  const port = store.controllers?.port
  const displayOn = (cfg: any) =>
    Object.keys(L.omitBy(cfg, (v) => !v))
      .map((v) => setting(v))
      .join(', ')
  const banner = `

  ${cyanBright(`â–„                             â–„             â–„ â–„`)}
  ${cyan(`â–ˆâ”€â–„  â–„ â–„  â–„â”€â–„  â–„â”€â–„  â–„â”€â–„  â–„â”€â”€  â–ˆâ”€   â”€â–„  â–„â”€â”€  â–ˆâ”€â–„`)}
  ${magentaBright(`â–€ â–€  â–€â”€â–ˆ  â–ˆâ”€â–€  â–€â”€â”€  â–€    â”€â”€â–€  â–€â”€  â–€â”€â–€  â–€â”€â”€  â–€ â–€`)}
  ${magentaBright(`     â””â”€â–€${gray(`              https://hyperstackjs.io`)}`)}


ðŸš€ App is running on port ${stat(port)}, in ${stat(store.app?.mode)} mode. 
ðŸ¤˜ Loaded ${stat(store.controllers?.numControllers)} controllers, ${stat(
    store.workers?.numWorkers
  )} workers, ${stat(store.workers?.numMailers)} mailers in ${magenta(
    parseInt(store.stats?.bootTime as any)
  )}ms.

ðŸ§° Environment is environments/${stat(store.app?.environmentFile)}
   Â· logger: ${setting(logger.level)}
   Â· controllers: ${displayOn(context.config().controllers)}
   Â· workers: ${setting(store.workers?.background)}, ${setting(
    store.workers?.mode
  )}
   Â· models:  ${displayOn(context.config().database)}
   Â· mailers: ${displayOn(
     L.omit(context.config().mailers, 'delivery')
   )} delivery=${setting(context.config().mailers?.delivery || 'none')}
`
  print(banner)
}
