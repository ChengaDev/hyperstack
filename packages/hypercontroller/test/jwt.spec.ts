import request from 'supertest'
import { sign } from 'jsonwebtoken'

import {
  Controller,
  Get,
  HttpResponseOK,
  authWithJWT,
  createServer,
} from '../src/index'
import { expectWithSnapshot, logger } from './utils'

const signJWT = (payload: any) =>
  sign(payload, 'testing-secret', {
    expiresIn: '1h',
  })

const { MustAuthWithJWT } = authWithJWT({
  secret: 'testing-secret',
  loader: async (payload: any) => ({
    user: { name: 'joe', email: payload.email },
  }),
})

@MustAuthWithJWT
@Controller('api')
class Api {
  @Get('foobar')
  async foobar(_req: any, _res: any) {
    return new HttpResponseOK({ its: 'alright' })
  }
}

const app = createServer({
  controllers: [Api],
  opts: { logging: { logger } },
})
describe('hypercontroller/jwt', () => {
  it('passes auth', async () => {
    const key = signJWT({ email: 'joe@ericsson.com' })
    await expectWithSnapshot(
      200,
      request(app)
        .get('/api/foobar')
        .set({ Authorization: `Bearer ${key}` })
    )
  })
  it('fails auth', async () => {
    await expectWithSnapshot(
      401,
      request(app).get('/api/foobar').set({ Authorization: 'Bearer bad-token' })
    )
    await expectWithSnapshot(401, request(app).get('/api/foobar'))
  })
})
