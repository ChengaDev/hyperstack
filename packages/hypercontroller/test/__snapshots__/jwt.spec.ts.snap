// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hypercontroller/jwt fails auth 1`] = `
Object {
  "body": Object {},
  "headers": Object {
    "connection": "close",
    "content-length": "9",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "text/html; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "www-authenticate": "",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 401,
}
`;

exports[`hypercontroller/jwt fails auth 2`] = `
Object {
  "body": Object {},
  "headers": Object {
    "connection": "close",
    "content-length": "8",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "text/html; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "www-authenticate": "",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 401,
}
`;

exports[`hypercontroller/jwt passes auth 1`] = `
Object {
  "body": Object {
    "its": "alright",
  },
  "headers": Object {
    "connection": "close",
    "content-length": "17",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;
